// agentmail/dist/cjs/serialization/resources/drafts/types/CreateDraftRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateDraftRequest = void 0;
const core = __importStar(require("../../../../core/index.js"));
const SendAttachment_js_1 = require("../../attachments/types/SendAttachment.js");
const DraftBcc_js_1 = require("./DraftBcc.js");
const DraftCc_js_1 = require("./DraftCc.js");
const DraftClientId_js_1 = require("./DraftClientId.js");
const DraftHtml_js_1 = require("./DraftHtml.js");
const DraftInReplyTo_js_1 = require("./DraftInReplyTo.js");
const DraftLabels_js_1 = require("./DraftLabels.js");
const DraftReplyTo_js_1 = require("./DraftReplyTo.js");
const DraftSendAt_js_1 = require("./DraftSendAt.js");
const DraftSubject_js_1 = require("./DraftSubject.js");
const DraftText_js_1 = require("./DraftText.js");
const DraftTo_js_1 = require("./DraftTo.js");
exports.CreateDraftRequest = core.serialization.object({
    labels: DraftLabels_js_1.DraftLabels.optional(),
    replyTo: core.serialization.property("reply_to", DraftReplyTo_js_1.DraftReplyTo.optional()),
    to: DraftTo_js_1.DraftTo.optional(),
    cc: DraftCc_js_1.DraftCc.optional(),
    bcc: DraftBcc_js_1.DraftBcc.optional(),
    subject: DraftSubject_js_1.DraftSubject.optional(),
    text: DraftText_js_1.DraftText.optional(),
    html: DraftHtml_js_1.DraftHtml.optional(),
    attachments: core.serialization.list(SendAttachment_js_1.SendAttachment).optional(),
    inReplyTo: core.serialization.property("in_reply_to", DraftInReplyTo_js_1.DraftInReplyTo.optional()),
    sendAt: core.serialization.property("send_at", DraftSendAt_js_1.DraftSendAt.optional()),
    clientId: core.serialization.property("client_id", DraftClientId_js_1.DraftClientId.optional()),
});
