// agentmail/dist/cjs/serialization/resources/drafts/types/DraftItem.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DraftItem = void 0;
const core = __importStar(require("../../../../core/index.js"));
const InboxId_js_1 = require("../../inboxes/types/InboxId.js");
const ThreadId_js_1 = require("../../threads/types/ThreadId.js");
const DraftAttachments_js_1 = require("./DraftAttachments.js");
const DraftBcc_js_1 = require("./DraftBcc.js");
const DraftCc_js_1 = require("./DraftCc.js");
const DraftId_js_1 = require("./DraftId.js");
const DraftLabels_js_1 = require("./DraftLabels.js");
const DraftPreview_js_1 = require("./DraftPreview.js");
const DraftSendAt_js_1 = require("./DraftSendAt.js");
const DraftSendStatus_js_1 = require("./DraftSendStatus.js");
const DraftSubject_js_1 = require("./DraftSubject.js");
const DraftTo_js_1 = require("./DraftTo.js");
const DraftUpdatedAt_js_1 = require("./DraftUpdatedAt.js");
exports.DraftItem = core.serialization.object({
    inboxId: core.serialization.property("inbox_id", InboxId_js_1.InboxId),
    threadId: core.serialization.property("thread_id", ThreadId_js_1.ThreadId),
    draftId: core.serialization.property("draft_id", DraftId_js_1.DraftId),
    labels: DraftLabels_js_1.DraftLabels,
    to: DraftTo_js_1.DraftTo.optional(),
    cc: DraftCc_js_1.DraftCc.optional(),
    bcc: DraftBcc_js_1.DraftBcc.optional(),
    subject: DraftSubject_js_1.DraftSubject.optional(),
    preview: DraftPreview_js_1.DraftPreview.optional(),
    attachments: DraftAttachments_js_1.DraftAttachments.optional(),
    sendStatus: core.serialization.property("send_status", DraftSendStatus_js_1.DraftSendStatus.optional()),
    sendAt: core.serialization.property("send_at", DraftSendAt_js_1.DraftSendAt.optional()),
    updatedAt: core.serialization.property("updated_at", DraftUpdatedAt_js_1.DraftUpdatedAt),
});
