// agentmail/dist/cjs/api/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.websockets = exports.webhooks = exports.threads = exports.pods = exports.organizations = exports.metrics = exports.messages = exports.inboxes = exports.events = exports.drafts = exports.domains = exports.attachments = exports.apiKeys = void 0;
__exportStar(require("./apiKeys/client/requests/index.js"), exports);
exports.apiKeys = __importStar(require("./apiKeys/index.js"));
__exportStar(require("./apiKeys/types/index.js"), exports);
exports.attachments = __importStar(require("./attachments/index.js"));
__exportStar(require("./attachments/types/index.js"), exports);
__exportStar(require("./domains/client/requests/index.js"), exports);
exports.domains = __importStar(require("./domains/index.js"));
__exportStar(require("./domains/types/index.js"), exports);
__exportStar(require("./drafts/client/requests/index.js"), exports);
exports.drafts = __importStar(require("./drafts/index.js"));
__exportStar(require("./drafts/types/index.js"), exports);
exports.events = __importStar(require("./events/index.js"));
__exportStar(require("./events/types/index.js"), exports);
exports.inboxes = __importStar(require("./inboxes/index.js"));
__exportStar(require("./messages/errors/index.js"), exports);
exports.messages = __importStar(require("./messages/index.js"));
__exportStar(require("./messages/types/index.js"), exports);
__exportStar(require("./metrics/client/requests/index.js"), exports);
exports.metrics = __importStar(require("./metrics/index.js"));
__exportStar(require("./metrics/types/index.js"), exports);
exports.organizations = __importStar(require("./organizations/index.js"));
__exportStar(require("./organizations/types/index.js"), exports);
exports.pods = __importStar(require("./pods/index.js"));
__exportStar(require("./threads/client/requests/index.js"), exports);
exports.threads = __importStar(require("./threads/index.js"));
__exportStar(require("./threads/types/index.js"), exports);
exports.webhooks = __importStar(require("./webhooks/index.js"));
exports.websockets = __importStar(require("./websockets/index.js"));
__exportStar(require("./websockets/types/index.js"), exports);
