// agentmail/dist/esm/serialization/resources/attachments/types/Attachment.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
import { AttachmentContentDisposition } from "./AttachmentContentDisposition.mjs";
import { AttachmentContentId } from "./AttachmentContentId.mjs";
import { AttachmentContentType } from "./AttachmentContentType.mjs";
import { AttachmentFilename } from "./AttachmentFilename.mjs";
import { AttachmentId } from "./AttachmentId.mjs";
import { AttachmentSize } from "./AttachmentSize.mjs";
export const Attachment = core.serialization.object({
    attachmentId: core.serialization.property("attachment_id", AttachmentId),
    filename: AttachmentFilename.optional(),
    size: AttachmentSize,
    contentType: core.serialization.property("content_type", AttachmentContentType.optional()),
    contentDisposition: core.serialization.property("content_disposition", AttachmentContentDisposition.optional()),
    contentId: core.serialization.property("content_id", AttachmentContentId.optional()),
});
