// agentmail/dist/esm/serialization/resources/drafts/types/CreateDraftRequest.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
import { SendAttachment } from "../../attachments/types/SendAttachment.mjs";
import { DraftBcc } from "./DraftBcc.mjs";
import { DraftCc } from "./DraftCc.mjs";
import { DraftClientId } from "./DraftClientId.mjs";
import { DraftHtml } from "./DraftHtml.mjs";
import { DraftInReplyTo } from "./DraftInReplyTo.mjs";
import { DraftLabels } from "./DraftLabels.mjs";
import { DraftReplyTo } from "./DraftReplyTo.mjs";
import { DraftSendAt } from "./DraftSendAt.mjs";
import { DraftSubject } from "./DraftSubject.mjs";
import { DraftText } from "./DraftText.mjs";
import { DraftTo } from "./DraftTo.mjs";
export const CreateDraftRequest = core.serialization.object({
    labels: DraftLabels.optional(),
    replyTo: core.serialization.property("reply_to", DraftReplyTo.optional()),
    to: DraftTo.optional(),
    cc: DraftCc.optional(),
    bcc: DraftBcc.optional(),
    subject: DraftSubject.optional(),
    text: DraftText.optional(),
    html: DraftHtml.optional(),
    attachments: core.serialization.list(SendAttachment).optional(),
    inReplyTo: core.serialization.property("in_reply_to", DraftInReplyTo.optional()),
    sendAt: core.serialization.property("send_at", DraftSendAt.optional()),
    clientId: core.serialization.property("client_id", DraftClientId.optional()),
});
