// agentmail/dist/esm/serialization/resources/drafts/types/DraftItem.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
import { InboxId } from "../../inboxes/types/InboxId.mjs";
import { ThreadId } from "../../threads/types/ThreadId.mjs";
import { DraftAttachments } from "./DraftAttachments.mjs";
import { DraftBcc } from "./DraftBcc.mjs";
import { DraftCc } from "./DraftCc.mjs";
import { DraftId } from "./DraftId.mjs";
import { DraftLabels } from "./DraftLabels.mjs";
import { DraftPreview } from "./DraftPreview.mjs";
import { DraftSendAt } from "./DraftSendAt.mjs";
import { DraftSendStatus } from "./DraftSendStatus.mjs";
import { DraftSubject } from "./DraftSubject.mjs";
import { DraftTo } from "./DraftTo.mjs";
import { DraftUpdatedAt } from "./DraftUpdatedAt.mjs";
export const DraftItem = core.serialization.object({
    inboxId: core.serialization.property("inbox_id", InboxId),
    threadId: core.serialization.property("thread_id", ThreadId),
    draftId: core.serialization.property("draft_id", DraftId),
    labels: DraftLabels,
    to: DraftTo.optional(),
    cc: DraftCc.optional(),
    bcc: DraftBcc.optional(),
    subject: DraftSubject.optional(),
    preview: DraftPreview.optional(),
    attachments: DraftAttachments.optional(),
    sendStatus: core.serialization.property("send_status", DraftSendStatus.optional()),
    sendAt: core.serialization.property("send_at", DraftSendAt.optional()),
    updatedAt: core.serialization.property("updated_at", DraftUpdatedAt),
});
