// agentmail/dist/cjs/serialization/resources/webhooks/types/CreateWebhookRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateWebhookRequest = void 0;
const core = __importStar(require("../../../../core/index.js"));
const EventTypes_js_1 = require("../../events/types/EventTypes.js");
const InboxIds_js_1 = require("../../events/types/InboxIds.js");
const PodIds_js_1 = require("../../events/types/PodIds.js");
const ClientId_js_1 = require("./ClientId.js");
const Url_js_1 = require("./Url.js");
exports.CreateWebhookRequest = core.serialization.object({
    url: Url_js_1.Url,
    eventTypes: core.serialization.property("event_types", EventTypes_js_1.EventTypes),
    podIds: core.serialization.property("pod_ids", PodIds_js_1.PodIds.optional()),
    inboxIds: core.serialization.property("inbox_ids", InboxIds_js_1.InboxIds.optional()),
    clientId: core.serialization.property("client_id", ClientId_js_1.ClientId.optional()),
});
