// agentmail/dist/cjs/serialization/resources/threads/types/Thread.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Thread = void 0;
const core = __importStar(require("../../../../core/index.js"));
const InboxId_js_1 = require("../../inboxes/types/InboxId.js");
const Message_js_1 = require("../../messages/types/Message.js");
const ThreadAttachments_js_1 = require("./ThreadAttachments.js");
const ThreadCreatedAt_js_1 = require("./ThreadCreatedAt.js");
const ThreadId_js_1 = require("./ThreadId.js");
const ThreadLabels_js_1 = require("./ThreadLabels.js");
const ThreadLastMessageId_js_1 = require("./ThreadLastMessageId.js");
const ThreadMessageCount_js_1 = require("./ThreadMessageCount.js");
const ThreadPreview_js_1 = require("./ThreadPreview.js");
const ThreadReceivedTimestamp_js_1 = require("./ThreadReceivedTimestamp.js");
const ThreadRecipients_js_1 = require("./ThreadRecipients.js");
const ThreadSenders_js_1 = require("./ThreadSenders.js");
const ThreadSentTimestamp_js_1 = require("./ThreadSentTimestamp.js");
const ThreadSize_js_1 = require("./ThreadSize.js");
const ThreadSubject_js_1 = require("./ThreadSubject.js");
const ThreadTimestamp_js_1 = require("./ThreadTimestamp.js");
const ThreadUpdatedAt_js_1 = require("./ThreadUpdatedAt.js");
exports.Thread = core.serialization.object({
    inboxId: core.serialization.property("inbox_id", InboxId_js_1.InboxId),
    threadId: core.serialization.property("thread_id", ThreadId_js_1.ThreadId),
    labels: ThreadLabels_js_1.ThreadLabels,
    timestamp: ThreadTimestamp_js_1.ThreadTimestamp,
    receivedTimestamp: core.serialization.property("received_timestamp", ThreadReceivedTimestamp_js_1.ThreadReceivedTimestamp.optional()),
    sentTimestamp: core.serialization.property("sent_timestamp", ThreadSentTimestamp_js_1.ThreadSentTimestamp.optional()),
    senders: ThreadSenders_js_1.ThreadSenders,
    recipients: ThreadRecipients_js_1.ThreadRecipients,
    subject: ThreadSubject_js_1.ThreadSubject.optional(),
    preview: ThreadPreview_js_1.ThreadPreview.optional(),
    attachments: ThreadAttachments_js_1.ThreadAttachments.optional(),
    lastMessageId: core.serialization.property("last_message_id", ThreadLastMessageId_js_1.ThreadLastMessageId),
    messageCount: core.serialization.property("message_count", ThreadMessageCount_js_1.ThreadMessageCount),
    size: ThreadSize_js_1.ThreadSize,
    updatedAt: core.serialization.property("updated_at", ThreadUpdatedAt_js_1.ThreadUpdatedAt),
    createdAt: core.serialization.property("created_at", ThreadCreatedAt_js_1.ThreadCreatedAt),
    messages: core.serialization.list(Message_js_1.Message),
});
