// agentmail/dist/esm/serialization/resources/messages/types/Message.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
import { InboxId } from "../../inboxes/types/InboxId.mjs";
import { ThreadId } from "../../threads/types/ThreadId.mjs";
import { MessageAttachments } from "./MessageAttachments.mjs";
import { MessageBcc } from "./MessageBcc.mjs";
import { MessageCc } from "./MessageCc.mjs";
import { MessageCreatedAt } from "./MessageCreatedAt.mjs";
import { MessageFrom } from "./MessageFrom.mjs";
import { MessageHeaders } from "./MessageHeaders.mjs";
import { MessageHtml } from "./MessageHtml.mjs";
import { MessageId } from "./MessageId.mjs";
import { MessageInReplyTo } from "./MessageInReplyTo.mjs";
import { MessageLabels } from "./MessageLabels.mjs";
import { MessagePreview } from "./MessagePreview.mjs";
import { MessageReferences } from "./MessageReferences.mjs";
import { MessageSize } from "./MessageSize.mjs";
import { MessageSubject } from "./MessageSubject.mjs";
import { MessageText } from "./MessageText.mjs";
import { MessageTimestamp } from "./MessageTimestamp.mjs";
import { MessageTo } from "./MessageTo.mjs";
import { MessageUpdatedAt } from "./MessageUpdatedAt.mjs";
export const Message = core.serialization.object({
    inboxId: core.serialization.property("inbox_id", InboxId),
    threadId: core.serialization.property("thread_id", ThreadId),
    messageId: core.serialization.property("message_id", MessageId),
    labels: MessageLabels,
    timestamp: MessageTimestamp,
    from: MessageFrom,
    replyTo: core.serialization.property("reply_to", core.serialization.list(core.serialization.string()).optional()),
    to: MessageTo,
    cc: MessageCc.optional(),
    bcc: MessageBcc.optional(),
    subject: MessageSubject.optional(),
    preview: MessagePreview.optional(),
    text: MessageText.optional(),
    html: MessageHtml.optional(),
    extractedText: core.serialization.property("extracted_text", core.serialization.string().optional()),
    extractedHtml: core.serialization.property("extracted_html", core.serialization.string().optional()),
    attachments: MessageAttachments.optional(),
    inReplyTo: core.serialization.property("in_reply_to", MessageInReplyTo.optional()),
    references: MessageReferences.optional(),
    headers: MessageHeaders.optional(),
    size: MessageSize,
    updatedAt: core.serialization.property("updated_at", MessageUpdatedAt),
    createdAt: core.serialization.property("created_at", MessageCreatedAt),
});
