// agentmail/dist/esm/serialization/resources/messages/types/SendMessageRequest.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
import { MessageHtml } from "./MessageHtml.mjs";
import { MessageLabels } from "./MessageLabels.mjs";
import { MessageSubject } from "./MessageSubject.mjs";
import { MessageText } from "./MessageText.mjs";
import { SendMessageAttachments } from "./SendMessageAttachments.mjs";
import { SendMessageBcc } from "./SendMessageBcc.mjs";
import { SendMessageCc } from "./SendMessageCc.mjs";
import { SendMessageHeaders } from "./SendMessageHeaders.mjs";
import { SendMessageReplyTo } from "./SendMessageReplyTo.mjs";
import { SendMessageTo } from "./SendMessageTo.mjs";
export const SendMessageRequest = core.serialization.object({
    labels: MessageLabels.optional(),
    replyTo: core.serialization.property("reply_to", SendMessageReplyTo.optional()),
    to: SendMessageTo.optional(),
    cc: SendMessageCc.optional(),
    bcc: SendMessageBcc.optional(),
    subject: MessageSubject.optional(),
    text: MessageText.optional(),
    html: MessageHtml.optional(),
    attachments: SendMessageAttachments.optional(),
    headers: SendMessageHeaders.optional(),
});
