// agentmail/dist/cjs/serialization/resources/messages/types/Message.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
const core = __importStar(require("../../../../core/index.js"));
const InboxId_js_1 = require("../../inboxes/types/InboxId.js");
const ThreadId_js_1 = require("../../threads/types/ThreadId.js");
const MessageAttachments_js_1 = require("./MessageAttachments.js");
const MessageBcc_js_1 = require("./MessageBcc.js");
const MessageCc_js_1 = require("./MessageCc.js");
const MessageCreatedAt_js_1 = require("./MessageCreatedAt.js");
const MessageFrom_js_1 = require("./MessageFrom.js");
const MessageHeaders_js_1 = require("./MessageHeaders.js");
const MessageHtml_js_1 = require("./MessageHtml.js");
const MessageId_js_1 = require("./MessageId.js");
const MessageInReplyTo_js_1 = require("./MessageInReplyTo.js");
const MessageLabels_js_1 = require("./MessageLabels.js");
const MessagePreview_js_1 = require("./MessagePreview.js");
const MessageReferences_js_1 = require("./MessageReferences.js");
const MessageSize_js_1 = require("./MessageSize.js");
const MessageSubject_js_1 = require("./MessageSubject.js");
const MessageText_js_1 = require("./MessageText.js");
const MessageTimestamp_js_1 = require("./MessageTimestamp.js");
const MessageTo_js_1 = require("./MessageTo.js");
const MessageUpdatedAt_js_1 = require("./MessageUpdatedAt.js");
exports.Message = core.serialization.object({
    inboxId: core.serialization.property("inbox_id", InboxId_js_1.InboxId),
    threadId: core.serialization.property("thread_id", ThreadId_js_1.ThreadId),
    messageId: core.serialization.property("message_id", MessageId_js_1.MessageId),
    labels: MessageLabels_js_1.MessageLabels,
    timestamp: MessageTimestamp_js_1.MessageTimestamp,
    from: MessageFrom_js_1.MessageFrom,
    replyTo: core.serialization.property("reply_to", core.serialization.list(core.serialization.string()).optional()),
    to: MessageTo_js_1.MessageTo,
    cc: MessageCc_js_1.MessageCc.optional(),
    bcc: MessageBcc_js_1.MessageBcc.optional(),
    subject: MessageSubject_js_1.MessageSubject.optional(),
    preview: MessagePreview_js_1.MessagePreview.optional(),
    text: MessageText_js_1.MessageText.optional(),
    html: MessageHtml_js_1.MessageHtml.optional(),
    extractedText: core.serialization.property("extracted_text", core.serialization.string().optional()),
    extractedHtml: core.serialization.property("extracted_html", core.serialization.string().optional()),
    attachments: MessageAttachments_js_1.MessageAttachments.optional(),
    inReplyTo: core.serialization.property("in_reply_to", MessageInReplyTo_js_1.MessageInReplyTo.optional()),
    references: MessageReferences_js_1.MessageReferences.optional(),
    headers: MessageHeaders_js_1.MessageHeaders.optional(),
    size: MessageSize_js_1.MessageSize,
    updatedAt: core.serialization.property("updated_at", MessageUpdatedAt_js_1.MessageUpdatedAt),
    createdAt: core.serialization.property("created_at", MessageCreatedAt_js_1.MessageCreatedAt),
});
