// agentmail/dist/cjs/serialization/resources/messages/types/ReplyAllMessageRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReplyAllMessageRequest = void 0;
const core = __importStar(require("../../../../core/index.js"));
const MessageHtml_js_1 = require("./MessageHtml.js");
const MessageLabels_js_1 = require("./MessageLabels.js");
const MessageText_js_1 = require("./MessageText.js");
const SendMessageAttachments_js_1 = require("./SendMessageAttachments.js");
const SendMessageHeaders_js_1 = require("./SendMessageHeaders.js");
const SendMessageReplyTo_js_1 = require("./SendMessageReplyTo.js");
exports.ReplyAllMessageRequest = core.serialization.object({
    labels: MessageLabels_js_1.MessageLabels.optional(),
    replyTo: core.serialization.property("reply_to", SendMessageReplyTo_js_1.SendMessageReplyTo.optional()),
    text: MessageText_js_1.MessageText.optional(),
    html: MessageHtml_js_1.MessageHtml.optional(),
    attachments: SendMessageAttachments_js_1.SendMessageAttachments.optional(),
    headers: SendMessageHeaders_js_1.SendMessageHeaders.optional(),
});
