// agentmail/dist/esm/serialization/resources/domains/types/Domain.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
import { PodId } from "../../pods/types/PodId.mjs";
import { ClientId } from "./ClientId.mjs";
import { DomainId } from "./DomainId.mjs";
import { FeedbackEnabled } from "./FeedbackEnabled.mjs";
import { VerificationRecord } from "./VerificationRecord.mjs";
import { VerificationStatus } from "./VerificationStatus.mjs";
export const Domain = core.serialization.object({
    podId: core.serialization.property("pod_id", PodId.optional()),
    domainId: core.serialization.property("domain_id", DomainId),
    status: VerificationStatus,
    feedbackEnabled: core.serialization.property("feedback_enabled", FeedbackEnabled),
    records: core.serialization.list(VerificationRecord),
    clientId: core.serialization.property("client_id", ClientId.optional()),
    updatedAt: core.serialization.property("updated_at", core.serialization.date()),
    createdAt: core.serialization.property("created_at", core.serialization.date()),
});
