// agentmail/dist/esm/serialization/resources/webhooks/types/Webhook.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
import { EventTypes } from "../../events/types/EventTypes.mjs";
import { InboxIds } from "../../events/types/InboxIds.mjs";
import { PodIds } from "../../events/types/PodIds.mjs";
import { ClientId } from "./ClientId.mjs";
import { Url } from "./Url.mjs";
import { WebhookId } from "./WebhookId.mjs";
export const Webhook = core.serialization.object({
    webhookId: core.serialization.property("webhook_id", WebhookId),
    url: Url,
    eventTypes: core.serialization.property("event_types", EventTypes.optional()),
    podIds: core.serialization.property("pod_ids", PodIds.optional()),
    inboxIds: core.serialization.property("inbox_ids", InboxIds.optional()),
    secret: core.serialization.string(),
    enabled: core.serialization.boolean(),
    updatedAt: core.serialization.property("updated_at", core.serialization.date()),
    createdAt: core.serialization.property("created_at", core.serialization.date()),
    clientId: core.serialization.property("client_id", ClientId.optional()),
});
