// agentmail/dist/esm/serialization/resources/threads/types/Thread.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
import { InboxId } from "../../inboxes/types/InboxId.mjs";
import { Message } from "../../messages/types/Message.mjs";
import { ThreadAttachments } from "./ThreadAttachments.mjs";
import { ThreadCreatedAt } from "./ThreadCreatedAt.mjs";
import { ThreadId } from "./ThreadId.mjs";
import { ThreadLabels } from "./ThreadLabels.mjs";
import { ThreadLastMessageId } from "./ThreadLastMessageId.mjs";
import { ThreadMessageCount } from "./ThreadMessageCount.mjs";
import { ThreadPreview } from "./ThreadPreview.mjs";
import { ThreadReceivedTimestamp } from "./ThreadReceivedTimestamp.mjs";
import { ThreadRecipients } from "./ThreadRecipients.mjs";
import { ThreadSenders } from "./ThreadSenders.mjs";
import { ThreadSentTimestamp } from "./ThreadSentTimestamp.mjs";
import { ThreadSize } from "./ThreadSize.mjs";
import { ThreadSubject } from "./ThreadSubject.mjs";
import { ThreadTimestamp } from "./ThreadTimestamp.mjs";
import { ThreadUpdatedAt } from "./ThreadUpdatedAt.mjs";
export const Thread = core.serialization.object({
    inboxId: core.serialization.property("inbox_id", InboxId),
    threadId: core.serialization.property("thread_id", ThreadId),
    labels: ThreadLabels,
    timestamp: ThreadTimestamp,
    receivedTimestamp: core.serialization.property("received_timestamp", ThreadReceivedTimestamp.optional()),
    sentTimestamp: core.serialization.property("sent_timestamp", ThreadSentTimestamp.optional()),
    senders: ThreadSenders,
    recipients: ThreadRecipients,
    subject: ThreadSubject.optional(),
    preview: ThreadPreview.optional(),
    attachments: ThreadAttachments.optional(),
    lastMessageId: core.serialization.property("last_message_id", ThreadLastMessageId),
    messageCount: core.serialization.property("message_count", ThreadMessageCount),
    size: ThreadSize,
    updatedAt: core.serialization.property("updated_at", ThreadUpdatedAt),
    createdAt: core.serialization.property("created_at", ThreadCreatedAt),
    messages: core.serialization.list(Message),
});
